### PR TITLE
Add mobile data to WCTracker

### PR DESCRIPTION
**Pull request by joshbetz**
Adds a telemetry endpoint to record WCTracker data from the apps. We track each platform independently, so we can see which versions of the apps are used on a given site.

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Adds a telemetry endpoint that we can use from the mobile apps to add mobile data usage to WCTracker.

### How to test the changes in this Pull Request:

1. Send a POST request to the `/wp-json/wc-telemetry/tracker` endpoint with body: `{ "platform": "ios", "version": "1.2" }`
    * Example (requires [the basic auth plugin](https://github.com/WP-API/Basic-Auth)): `curl -u user:password -X POST -H "Content-Type: application/json" -d '{ "platform": "ios", "version": "1.2" }' http://localhost/wp-json/wc-telemetry/tracker`
2. Via WP-CLI or similar, execute `WC_Tracker::get_tracking_data()`. Observe that `wc_mobile_usage` shows the correct platform and version.
    * Example:  `wp eval 'var_dump(WC_Tracker::get_tracking_data());'`
3. Send another POST request to the telemetry endpoint with body: `{ "platform": "ios", "version": "1.3" }`
4. Observe that `wc_mobile_usage` has recorded the updated version correctly.
5. Send another POST request to the telemetry endpoint with the original body: `{ "platform": "ios", "version": "1.2" }`
6. Observe that `wc_mobile_usage` has NOT updated with the older version number.

Note: If using `wp shell` to test, you may have to close the shell session and re-open if the option value is cached.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Dev - Add mobile data to WCTracker

### FOR PR REVIEWER ONLY:

* [x] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: includes/rest-api/Controllers/Version2/class-wc-rest-telemetry-v2-controller.php [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+	/**
+	 * Route base.
+	 *
+	 * @var string
+	 */
+	protected $rest_base = 'telemetry';
+
+	/**
+	 * Register the route for /system_status
```
Comment (jonathansadowski) [inline includes/rest-api/Controllers/Version2/class-wc-rest-telemetry-v2-controller.php]: Looks like the comment here was carried over from the class used as an example.

File: includes/rest-api/Controllers/Version3/class-wc-rest-telemetry-controller.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
+<?php
+/**
+ * REST API WC Telemetry Controller
+ *
+ * Handles requests to the /system_status/App/* endpoints.
```
Comment (jonathansadowski) [inline includes/rest-api/Controllers/Version3/class-wc-rest-telemetry-controller.php]: This docblock doesn't seem to be accurate.

**Review by jonathansadowski — CHANGES_REQUESTED** [review]
Sorry for the delay in reviewing this. I commented about some comments/docblocks that seem to be inaccurate / need updating, otherwise this tested well for me.

I do have some inarticulable pause about adding telemetry to the v3 namespace. I do wonder if it makes sense to place it in some other namespace. Placing it there really encourages its usage outside of the apps, in my mind... but I'm not sure if that's just me. /cc @vedanshujain did you have any opinions in that regard?

**Comment by joshbetz** [comment]
Thanks @jonathansadowski. I'll double check all the comment blocks and make sure to fix the ones you mentioned.

> I do have some inarticulable pause about adding telemetry to the v3 namespace.

I don't have any strong opinions on this. Happy to put it in a new namespace or even `wc-analytics`.

File: tests/e2e/core-tests/specs/api/telemetry.test.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+		} );
+
+		it.each([
+			null,
+			{},
+			{ platform: 'ios' },
+			{ version: '1.1' },
+		])( 'errors for invalid request body - %p', async data => {
+			const response = await client
+				.post( `/wc/v3/telemetry`, data )
```
Comment (jonathansadowski) [inline tests/e2e/core-tests/specs/api/telemetry.test.js]: Looks like this still needs to be updated with the new route

File: includes/rest-api/Controllers/Telemetry/class-wc-rest-telemetry-controller.php [hunk 4]
```diff
@@ -0,0 +1,142 @@
+<?php
+/**
+ * REST API WC Telemetry controller
+ *
+ * Handles requests to the /telemetry endpoint.
```
Comment (jonathansadowski) [inline includes/rest-api/Controllers/Telemetry/class-wc-rest-telemetry-controller.php]: Can you update this to `/tracker`?

File: includes/rest-api/Controllers/Telemetry/class-wc-rest-telemetry-controller.php [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+	/**
+	 * Route base.
+	 *
+	 * @var string
+	 */
+	protected $rest_base = 'tracker';
+
+	/**
+	 * Register the route for /telemetry
```
Comment (jonathansadowski) [inline includes/rest-api/Controllers/Telemetry/class-wc-rest-telemetry-controller.php]: Same as above

File: tests/e2e/core-tests/specs/api/telemetry.test.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+				.catch( err => {
+					expect( err.statusCode ).toBe( 400 );
+				} );
+
+			expect( response ).toBeUndefined();
+		} );
+
+		it( 'returns 200 with correct fields', async () => {
+			const response = await client
+				.post( `/wc/v3/telemetry`, {
```
Comment (jonathansadowski) [inline tests/e2e/core-tests/specs/api/telemetry.test.js]: Same as above

**Review by jonathansadowski — CHANGES_REQUESTED** [review]
Looking good — I added a few minor comments, but otherwise everything tested out well. Thanks for updating the namespace used!

**Review by jonathansadowski — DISMISSED** [review]
Thanks for making the requested updates. Everything looks good. I verified that the test failures are unrelated to this change, and that the unit tests pass for me when I run locally with this branch.

**Review by jonathansadowski — APPROVED** [review]
Thanks for the rebase. All tests good except for the sniff re: `Server.php` and the filename, but I don't think we should fix that in the context of this PR.

Looking good, nice work.

**Comment by jonathansadowski** [comment]
I didn't add the needs changelog or needs testing instructions labels because of the scope of this API being limited to mobile consumption.

**Comment by Konamiman** [comment]
I have updated the testing instructions since the endpoint is actually `/wp-json/wc-telemetry/tracker`, not `/wp-json/wc/v3/telemetry` (and I have added `curl` and WP CLI examples too).

**Pull request merged**